scipy/stats/_sobol.pyx

```diff
@@ -0,0 +1,292 @@
+from __future__ import division, absolute_import
+
+cimport cython
+cimport numpy as cnp
+
+import os
+import math
+import numpy as np
+from scipy.stats import norm
+
+cdef int MAXDIM = 21201  # max number of dimensions
+cdef int MAXDEG = 18  # max polynomial degree
+cdef int MAXBIT = 30  # max number of bits
+
+cdef int poly[21201]
+cdef int vinit[21201][18]
+
+cdef int LARGEST_NUMBER = 2 ** MAXBIT  # largest possible integer
+cdef float RECIPD = 1.0 / LARGEST_NUMBER  # normalization constant
+
+
+# Load direction numbers (taken from https://web.maths.unsw.edu.au/~fkuo/sobol/)
+# TODO: Lazy loading
```

Comment:
Not quite sure how to best do this - are there best practices for this in scipy? The main thing we need to avoid is that this data file is loaded on every import. If we're a little more ambitious we could try to lazy-load this on the first Sobol generation.

Reply:
I could not find anything in SciPy for that. I will try to propose something which loads it in the python code in `Sobol` at the first instantiation only (through a metaclass or a lazy property?? something like this) and then we can pass the lists along. But doing it will involve moving around more data. Otherwise, I have no clue how to do this only in cython. Will try to find something.

Reply:
it may be as simple as having an `initialize` method in the Cython code that gets called in the python `Sobol` code and loads the data into the arrays. Then we could have some kind of `is_initialized` variable in the cython code that shortcuts the loading in case it has been done before. Not entirely sure whether that would cause any scoping issues though.

Reply:
This seems to be working:

```python
cdef bint is_initialized = False

# Load direction numbers (taken from https://web.maths.unsw.edu.au/~fkuo/sobol/)
def initialize():
    dns = np.load(os.path.join(os.path.dirname(__file__), "_sobol_direction_numbers.npz"))
    print('toto')
    poly = dns["poly"]
    vinit = dns["vinit"]
    is_initialized = True


if not is_initialized:
    initialize()
```

```python
>>> from scipy.stats.qmc import Sobol
toto
>>> Sobol(2)
<scipy.stats.qmc.Sobol object at 0x7fe23749dc70>
>>> from scipy.stats.qmc import Sobol
>>> from scipy.stats.qmc import Sobol
>>> engine = Sobol(2)
>>> engine.random()
array([[0.5, 0. ]])
```

Reply:
Ah ok no I see: I need an extra step to not trigger this with `import scipy.stats.qmc`

Reply:
This is working:

```python
cdef bint is_initialized = False

# Load direction numbers (taken from https://web.maths.unsw.edu.au/~fkuo/sobol/)
def initialize_direction_numbers():
    global is_initialized
    if not is_initialized:
        dns = np.load(os.path.join(os.path.dirname(__file__), "_sobol_direction_numbers.npz"))
        print('toto')
        poly = dns["poly"]
        vinit = dns["vinit"]
        is_initialized = True
```

And I need a call in `Sobol `to `initialize_direction_numbers ` in the init.

```python
>>> import scipy.stats.qmc as qmc
>>> engine = qmc.Sobol(2)
toto
>>> engine2 = qmc.Sobol(2)

>>>
```

Reply:
Nice, that looks great.

Note that I haven't done any real validation of the behavior with the new 21201-dim direction numbers. Before we land this we'll need to do some due diligence (and update the tests).

Reply:
If we really wanted to be ambitious we could allow the user to choose from the direction numbers obtained using the other search criteria. But that seems like something we can add as a feature down the road.

Reply:
I am looking into the tests right now. So far, just values to change. I will about how to "verify" this more.

We could add other files I guess, or even give the possibility to provide your file. But this is really an advance option. As you said, maybe we just move this to another feature. I would even say that if we say we "cannot yet" verify the new directions that we can also make it as another feature.

Reply:
Mmm all normality checks are falling now.

Reply:
Ok there is an issue with the generation. Only the first dimension is getting something else than 0.

```python
engine = qmc.Sobol(3)
sample = engine.random(250)
sample
array([[0.5       , 0.        , 0.        ],
       [0.75      , 0.        , 0.        ],
       [0.25      , 0.        , 0.        ],
...
```

Reply:
mhhh my hunch is that the `poly` and `vinit` arrays are actually not initialized properly with the lazy loading setup. That's where I'd start debugging

Reply:
That was it 😉  I had to put both as global as well. The compiler was not complaining about it so I did not notice it.

Reply:
Cool. I think there is still some validation that needs to be done that the samples are ok statistically / discrepancy-wise. But good to know the lazy loading works properly.

Reply:
OK it's done! I am just looking into the star-discrepancy-L2. I am not sure we should include it. As said in Fang2006: 
> However, it has been shown that the star L2-discrepancy ignores differences [...] on any low-dimensional subspace and is not suitable for design of computer experiments.

But I don't mind having it. I am more looking into putting the wrap around discrepancy (I already have an optimized numpy version). (Careful, in Fang and more recent literature, there is an error on the equation. If you take the original articles of the Wrap around discrepancy you will see it).

Still on the discrepancy, I will remove the `bounds` parameter to be coherent with the rest of the API.

Appart from this, do you see anything else to do?

Reply:
Forgot to say that I also have a numpy version of the Mixture Discrepance (MD) which is described as more reasonable than both C2 and WD (https://reader.elsevier.com/reader/sd/pii/S0885064X12001057?token=197AF1FFBB724C86CD341789BC8FD9B8AB8BEB9D6B8F2B186C3B80E20A59493AE695209C7C21822B9239C43AF879037B)

Reply:
I have done a single function for the discrepancy and added a `method` param. This way, the user can decide between centered, wrap around, mixture and star L2 discrepancy.